### PR TITLE
Add decimal type support, and fix logical types for nested fields in Avro/Kafka.

### PR DESCRIPTION
**Pull request by jcferretti**

### CODE REVIEW

File: extensions/kafka/src/main/java/io/deephaven/kafka/ingest/GenericRecordChunkAdapter.java [hunk 1]
```diff
@@ -47,63 +50,73 @@ public static GenericRecordChunkAdapter make(
...
-                return new GenericRecordDoubleFieldCopier(fieldName);
+                return new GenericRecordDoubleFieldCopier(fieldPathStr, separator);
             case Object:
                 if (dataType == String.class) {
-                    return new GenericRecordStringFieldCopier(fieldName);
-                } else {
-                    return new GenericRecordObjectFieldCopier(fieldName);
+                    return new GenericRecordStringFieldCopier(fieldPathStr, separator);
+                }
+                final Schema.Field field = schema.getField(fieldPathStr);
```
Comment (rcaudy) [inline extensions/kafka/src/main/java/io/deephaven/kafka/ingest/GenericRecordChunkAdapter.java:107]: Do we still need this lookup or null check? I feel like this code should be more like the `DateTime` code:
```
                if (dataType == BigDecimal.class) {
                    final String[] fieldPath = GenericRecordUtil.getFieldPath(fieldPathStr, separator);
                    final Schema fieldSchema = GenericRecordUtil.getFieldSchema(schema, fieldPath);
                    final LogicalType logicalType = fieldSchema.getLogicalType();
                    if (logicalType instanceof LogicalTypes.Decimal) {
                        final LogicalTypes.Decimal decimalType = (LogicalTypes.Decimal) logicalType;
                        return new GenericRecordBigDecimalFieldCopier(
                                fieldPathStr, separator,
                                decimalType.getPrecision(), decimalType.getScale());
                    }
                    // Fall through to GenericRecordObjectFieldCopier? Or:
                    throw new IllegalArgumentException(
                            "Can not map field with unknown logical type to BigDecimal: field=" + fieldPathStr
                                    + ", type=" + logicalType);
                }
```

File: extensions/kafka/src/main/java/io/deephaven/kafka/ingest/GenericRecordUtil.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        while (i < fieldPath.length - 1) {
+            parentRecord = (GenericRecord) parentRecord.get(fieldPath[i]);
+            ++i;
+        }
+        return parentRecord.get(fieldPath[i]);
+    }
+
+    private static String buildPartialFieldName(final String[] fieldPath, final int k) {
+        final StringBuilder sb = new StringBuilder("(");
+        for (int i = 0; i < k; ++i) {
```
[the remark this answers is not in the document]
Comment (jcferretti) [reply extensions/kafka/src/main/java/io/deephaven/kafka/ingest/GenericRecordUtil.java:49]: Is gone tho, per your previous suggestion.

File: extensions/kafka/src/main/java/io/deephaven/kafka/ingest/GenericRecordUtil.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+     *        path to an individual desired field
+     * @return The value in {@code record} for the given {@code fieldPath}
+     */
+    public static Object getPath(final GenericRecord record, final String[] fieldPath) {
+        if (record == null) {
+            return null;
+        }
+        GenericRecord parentRecord = record;
+        int i = 0;
+        while (i < fieldPath.length - 1) {
```
[the remark this answers is not in the document]
Comment (jcferretti) [reply extensions/kafka/src/main/java/io/deephaven/kafka/ingest/GenericRecordUtil.java:70]: Very good catch, thanks, and done.

File: extensions/kafka/src/main/java/io/deephaven/kafka/ingest/GenericRecordUtil.java [hunk 4]
```diff
@@ -44,22 +74,31 @@ public static Object getPath(final GenericRecord record, final String[] fieldPat
...
-            if (i > 0) {
-                sb.append(", ");
-            }
-            sb.append(fieldPath[i]);
+    /**
+     * Get the value out of a Generic Record for a given field path (including nesting).
+     *
+     * @param record A {@code GenericRecord} from which to extract a value.
+     * @param fieldPath An array of individual field position indices, in order according to nesting from the root
+     *        representing repath to an individual desired field
```
Comment (rcaudy) [inline extensions/kafka/src/main/java/io/deephaven/kafka/ingest/GenericRecordUtil.java:82]: `repath`? Typo?

**Review by rcaudy — DISMISSED** [review]
Approving in case I was wrong about the typo.

**Review by rcaudy — APPROVED** [review]

**Pull request merged**